src/librustc_error_codes/error_codes/E0308.md

```diff
@@ -15,3 +15,47 @@ let x: i32 = "I am not a number!";
 //      |
 //    type `i32` assigned to variable `x`
 ```
+
+One possible scenario is that we have a trait and want to use multiple
+implementors conditionally via dynamic dispatch. E. g.
+
+```compile_fail,E0308
+use std::{io, fs, env};
+
+let arg = env::args().nth(1);
+
+let mut input = if arg == "-" {
+	std::io::stdin()
+} else {
+	fs::File::open(arg)?
+//	^^^^^^^^^^^^^^^^^^^^ expected struct `std::io::Stdin`, found struct `std::fs::File
+};
+// - if and else have incompatible types
+
+Ok::<(), io::Error>(())
+```
+
+In many cases, the trait methods use self by (mutable) reference. Thus
+we can create a `dyn` reference, but we need two stack slots to hold
```

Comment:
This feels sorta "out of nowhere", esp. the bits about stack slots, which doesn't feel particularly helpful. This feels too low level.

Reply:
A little rephrasing might be nice yes. And instead of talking about stacks, maybe say "to be sure that objects live enough" or something along the line?

Reply:
Good idea. Changed.